include/GafferScene/EditScopeAlgo.h

```diff
@@ -0,0 +1,83 @@
+//
+//  Copyright (c) 2019, Cinesite VFX Ltd. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERSCENE_EDITSCOPEALGO_H
+#define GAFFERSCENE_EDITSCOPEALGO_H
+
+#include "Gaffer/EditScope.h"
+#include "Gaffer/TransformPlug.h"
+
+#include "GafferScene/ScenePlug.h"
+
+namespace GafferScene
+{
+
+namespace EditScopeAlgo
+{
+
+// Pruning
+// =======
+
+GAFFERSCENE_API void setPruned( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path, bool pruned );
+GAFFERSCENE_API void setPruned( Gaffer::EditScope *scope, const IECore::PathMatcher &paths, bool pruned );
+GAFFERSCENE_API bool getPruned( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path );
+
+// Transforms
+// ==========
+//
+// These methods manipulate the local transform of a location.
+
+GAFFERSCENE_API void setTransform( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path, const Imath::M44f &matrix );
+GAFFERSCENE_API boost::optional<Imath::M44f> getTransform( const Gaffer::EditScope *scope, const ScenePlug::ScenePath &path );
+GAFFERSCENE_API bool hasTransform( const Gaffer::EditScope *scope, const ScenePlug::ScenePath &path );
+GAFFERSCENE_API void removeTransform( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path );
```

Comment:
Was it a conscious thing to make the shape of these two different edit tasks different? I think theres something bugging me about the asymmetry here between transform and prune. 

Transform has `remove`, `has` and `get`, but the `get` also returns an optional (which infers `has`). Prune wraps it all up with the inference that `get` returning false also means no edit is authored (which _is_ true), and no `remove` as setting it false is technically removing the edit. It feels like it would be good for there to be a consistency across all well-known algo defined edits. Always having `has*`/`remove*` would make more generic understanding of edits easier.

If we made `get` always return a `boost::optional` too, even for logically combined edits like prune, as it avoids one reason to have any first-class understanding of the meaning of the value of an edit, and could mean you could then ditch `has*` if you really wanted.

Maybe speculating too much (as I know its early days yet), but I think I'm currently leaning towards having a `has` method. I was just attempting to think about writing something like the HUD for a light tweak. We display the resulting in-scene value, and don't show the 'edit' values until the editor is opened. So to draw the panel, you'd just want to know `has*` for any given scope so you could decorate the presentation of the value.  When you then show the edit widget, I think we imagined that you'd then just be displaying a `PlugValueWidget` for the acquired plugs (so actually, thinking about it, we probably always want an `acquire*Plugs` too then!). The code that then helps surface edit status being full of `has` feels like it'd make it a bit more maintainable than `get`s that then don't really care about the result other than its not `None`? 

At the risk of drifting off into the unknown even further, I guess this surfacing of contributing edit scopes for a location is really part of the generic 'edit history' UI we've been discussing as the basis of Scene Inspector, light editors, etc... Having the standard suite of methods could allow a registry of edits for attributes/etc that maps to the algo functions rather than a lot of explicit lookups? Obviously this wouldn't attempt to facilitate editing itself, but could then go a long way to generalising where edits are and being able to remove them?

Reply:
> Was it a conscious thing to make the shape of these two different edit tasks different? I think theres something bugging me about the asymmetry here between transform and prune.

It was a conscious thing. The asymmetry bugged me a bit too, they're different operations and a one-size-fits-all approach doesn't work for them. The transform is a property of a location, it exists whether it has been edited or not. Pruning gains performance benefits from a batch API (`setPruned( PathMatcher )`) whereas transform edits don't. "Pruned" is not going to be a field in the HUD, because if it has been pruned, you can't select it.

That's not to say I'm not interested in a uniform API where possible; I am. It's just unfortunate that these are the first two things in EditScopeAlgo, and pruning doesn't match the pattern that we should try to establish for everything else, where we typically will be editing a property of some sort.

For what it's worth, I'm not wholly convinced by `setTransform/getTransform/hasTransform`. The only fundamental operations are `acquireTransformPlugs()` and `removeTransform()`, and I think only `acquireTransformPlugs()` is used in anger by any of the code I've written so far (upcoming PR). I'm also not 100% convinced by the `optional` returned by `getTransform()`; it means you need a separate way of querying the property when it hasn't been edited. But then again, chances are that in the HUD you need to be querying the property from the downstream viewed node, not from the EditScope. So maybe the `optional` is legit.

How this extends to something like a light tweak is a little bit unclear too. There presumably the `acquire()` method will return a `TweakPlug`, but it feels like `set` should maybe just do a replace? And what would `get` return? Are we going to introduce a struct of some sort to encode everything that a TweakPlug does? If we are, why don't we just leave it at `acquire()` and `remove()`, and then everything else goes through the dedicated plug API that is already designed for the task? The only reason I couldn't do that for transforms was the requirement that the TransformPlug be split into multiple columns for the spreadsheet, which is great for the user, but poor for the API.

Really I want to answer these questions later, so I can focus on getting the TransformTool integration over the line. I was half of a mind to just rip out `set/get/has` methods entirely, but figured (perhaps wrongly) that they were probably close enough, and that I could probably maintain compatibility if they needed changes later.

> I was just attempting to think about writing something like the HUD for a light tweak...At the risk of drifting off into the unknown even further...Having the standard suite of methods could allow a registry of edits for attributes/etc that maps to the algo functions rather than a lot of explicit lookups?

The HUD is going to have to interact with something one level up from EditScopeAlgo, and that's the thing that is going to have to present an entirely uniform API. At the EditScopeAlgo level, there are going to be several classes of things, each of which will require different arguments (`attributeName` for an attribute, `attributeName/shaderName/parameterName` for a shader parameter, etc).

Reply:
> "Pruned" is not going to be a field in the HUD, because if it has been pruned, you can't select it.

True indeed!